libmamba/src/solv-cpp/repo.hpp

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2023, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#ifndef MAMBA_SOLV_REPO_HPP
+#define MAMBA_SOLV_REPO_HPP
+
+#include <filesystem>
+#include <optional>
+#include <string_view>
+#include <utility>
+
+#include "solv-cpp/ids.hpp"
+#include "solv-cpp/solvable.hpp"
+
+typedef struct s_Repo Repo;
+
+namespace mamba::solv
+{
+
+    class ObjRepoViewConst
+    {
+    public:
+
+        explicit ObjRepoViewConst(const ::Repo* repo) noexcept;
+
+        auto raw() const -> const ::Repo*;
+
+        auto id() const -> RepoId;
+
+        auto name() const -> std::string_view;
+        auto url() const -> std::string_view;
+        auto channel() const -> std::string_view;
+        auto subdir() const -> std::string_view;
+
+        auto n_solvables() const -> std::size_t;
+        auto has_solvable(SolvableId id) const -> bool;
+        auto get_solvable(SolvableId id) const -> std::optional<ObjSolvableViewConst>;
+
+        template <typename UnaryFunc>
+        void for_each_solvable_id(UnaryFunc func) const;
+        template <typename UnaryFunc>
+        void for_each_solvable(UnaryFunc func) const;
+
+        /**
+         * Write repository information to file.
+         *
+         * @param solv_file A standard path with system encoding.
+         */
+        void write(std::filesystem::path solv_file) const;
+
+    private:
+
+        const ::Repo* m_repo = nullptr;
+    };
+
+    class ObjRepoView : public ObjRepoViewConst
```

Comment:
I forgot to add a comment here. I'm not in favor of mixing public inheritance and value semantic. A solution could be to have:
- `ObjRepoViewConstBase` and `ObjRepoViewBase` with constructors, destructors and assignment operators in the protected section.
- ObjRepoViewConst and ObjRepoView that prvately inherits from the previous bases, and "import" the method in thir public section.
 
This would avoid mixing / assigning objects of different types.

Reply:
Is this value semantic, not reference/view semantic? A copy still points to the same data, same as in `std::span`.

The two types here serve to represent the inner `const` of an `std::span`.
We do want a `ObjRepoView` to convert to an `ObjRepoViewConst`, so that one can call`f(ObjRepoViewConst)` with a  `ObjRepoView`.

There are no invalid operations in the current design, so IMHO using base types etc would be amount to reproducing what is already here.

Reply:
> Is this value semantic, not reference/view semantic?

As soon as the object is manipulated with its real type, and can be copied and/or moved, it is value semantic; the opposite is entity semantic, where you manipulate objects via pointers or reference to a common base type (and ideally you never want to know the real type after the instantiation).

If you want to authorize passing `ObjRepoView` to functions accepting `ObjReoiViewConst` then indeed the current design works, but keep in mind that this authorizes the following statements:

```cpp
ObjRepoView v(...);
ObjRepoViewConst cv = v; // (potentially partial copy, here it works because only the base class has data members)

ObjRepoView* v = new ObjRepoView(....);
void free_view(ObjRepoViewConst* cv)
{
    delete cv; // The destructor of ObjRepoView is not called since it is not virtual (and it should not be).
}
```

I agree this is not problematic here because there is only one data member in the base class, the destructor of `ObjRepoView` is trivial and the hierarchy has two classes only. It's just potentially dangerous regarding future changes, and a kind of weird to read when you are used to a clean distinction between those semantics.

Here is a detailed article that states this a bit differently (it is more about virtuality and entity semantics), epecially the second question about the destructor.

But again, if we really want to be able to bind `view` objects to `const view` references, then I'm not sure we can do it differently (with conversion constructors, we could bind to const references only).